notebooks/02-ingestion/02-batch-ingestion.ipynb

```diff
@@ -0,0 +1,1008 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6561d663-9cfe-4a62-80cc-dcbb0ce199ff",
+   "metadata": {},
+   "source": [
+    "#### Transform Data During Ingestion\n",
+    "\n",
+    "The SQL language provides a rich [set of functions](https://druid.apache.org/docs/latest/querying/sql-scalar.html) that can be applied to input columns to transform the data as it is being ingested. All scalar SQL function are available for normal ingestion. Rollup ingestion is discussed in the [Rollup Notebook](05-rollup.ipynb) which includes the use of aggregate functions at ingestion time as well.\n",
+    "\n",
+    "Here are some examples of such transformations:\n",
+    "\n",
+    "##### Time manipulation\n",
+    "There are many [time parsing and manipulation functions](https://druid.apache.org/docs/latest/querying/sql-scalar.html#date-and-time-functions) available in Apache Druid. It is common to do some time cleansing/transformation at ingestion. Here are some examples of time manipulation functions:\n",
+    "```\n",
+    "  TIME_PARSE( \"timestamp\") AS \"__time\",   \n",
+    "  TIME_FLOOR( TIME_PARSE( \"timestamp\"), 'P1W') AS \"week_start\",\n",
+    "  TIMESTAMPDIFF( DAY,\n",
+    "                 TIME_FLOOR( TIME_PARSE( \"timestamp\"), 'P1W'),\n",
+    "                 TIME_PARSE( \"timestamp\")\n",
+    "               ) AS \"days_since_week_start\"\n",
+    "   \n",
+    "```\n",
+    "\n",
+    "##### Use CASE statements to transform data\n",
+    "CASE statements can be used to resolve complex logic and prepare columns for certain query patterns. \n",
+    "Examples:\n",
+    "```\n",
+    "  CASE\n",
+    "     WHEN UPPER(\"adblock_list\")='NOADBLOCK' THEN 0\n",
+    "     ELSE 1\n",
+    "  END AS adblock_count,\n",
+    "\n",
+    "  CASE\n",
+    "     WHEN UPPER(\"adblock_list\")='EASYLIST' THEN 1\n",
+    "     ELSE 0\n",
+    "  END AS easylist_count\n",
+    "```\n",
+    "The two case statements above are examples of converting a categorical column like `adblock_list` into a numerical column that can be used as a meaningful metric when aggregated across different dimensions to get the count of events that were affected by an ad blocker.\n",
+    "\n",
+    "##### String manipulation\n",
+    "Apache Druid has [string manipulation functions](https://druid.apache.org/docs/latest/querying/sql-scalar.html#string-functions) that can be very useful for transformation during ingestion. Some examples:\n",
+    "```\n",
+    "  REPLACE(REGEXP_EXTRACT(\"app_version\", '[^\\.]*\\.'),'.','') AS major_version,\n",
+    "  STRING_TO_ARRAY(\"app_version\",'\\.') AS version_array,\n",
+    "  ARRAY_ORDINAL(STRING_TO_ARRAY(\"app_version\",'\\.'),3) AS patch_version\n",
+    "```\n",
+    "The above makes use of regex-based extraction, string replacement, string to array conversion and access to array elements as examples of the string transformation functions available.\n",
+    "\n",
+    "##### Data Flattening functions\n",
```

Comment:
Add link out to the nested columns notebook here.

Reply:
see comment above